### ISSUE TITLE
Parsing strings that look like numbers changed in release 0.87

### ISSUE THREAD

**Issue by veryrusty** [ingydotnet/yaml-libyaml-pm#111]
The change to turn off the internal POK flag in release 0.87 has impacted how strings that look like numbers are parsed.

Using a variation of the code from #68:
```
use v5.16;
use YAML::XS qw/ Load /;
use Devel::Peek qw/ Dump /;
use Data::Dumper qw/ Dumper /;
my $yaml = "foo: 0003\nbar: 03.30";
my $data = Load $yaml;
say Dumper $data;
Dump $data->{foo};
Dump $data->{bar};
```

YAML::XS@0.86 gives strings
```
$VAR1 = {
          'bar' => '03.30',
          'foo' => '0003'
        };

SV = PVIV(0x5648b041dde0) at 0x5648b04145f8
  REFCNT = 1
  FLAGS = (IOK,POK,pIOK,pPOK)
  IV = 3
  PV = 0x5648b0422f60 "0003"\0
  CUR = 4
  LEN = 10
SV = PVNV(0x5648b03fb300) at 0x5648b0414688
  REFCNT = 1
  FLAGS = (NOK,POK,pIOK,pNOK,pPOK)
  IV = 3
  NV = 3.3
  PV = 0x5648b05e4000 "03.30"\0
  CUR = 5
  LEN = 10
```

While YAML::XS@0.87 has munged them, numified `0003` and dropped the zeros from `03.30`
```
$VAR1 = {
          'bar' => '3.3',
          'foo' => 3
        };

SV = PVNV(0x55b424e98300) at 0x55b424eb15f8
  REFCNT = 1
  FLAGS = (IOK,pIOK)
  IV = 3
  NV = 3
  PV = 0x55b424ec00a0 "0003"\0
  CUR = 4
  LEN = 10
SV = PVNV(0x55b424e98320) at 0x55b424eb1688
  REFCNT = 1
  FLAGS = (NOK,pNOK)
  IV = 3
  NV = 3.3
  PV = 0x55b424fb2c10 "3.3"\0
  CUR = 3
  LEN = 32
```

With $YAML::XS::QuoteNumericStrings defaulting to true, I'd expect those leading zeros to be preserved (as the 0.86 release does).

**Comment by perlpunk** [comment]
First of all, `$YAML::XS::QuoteNumericStrings` is an option for dumping. It will quote strings that look like numbers, to make sure they are loaded as strings again.

The behaviour you are seeing is an incompatible change, I can see that.
However, it's more correct. An unquoted scalar `03.30` or `0003` is a number according to the YAML spec. The zeroes are presentation details that should not be preserved.
Can you tell us more about the use case that relies on this behaviour?

**Comment by veryrusty** [comment]
> An unquoted scalar is a number according to the YAML spec

Double checking I've understood the spec correctly; that arrises from the extended list of regexs from the core schema?

> Can you tell us more about the use case that relies on this behaviour

We have 18+ years of data collection specifications that use YAML docs as their primary source. They get processed into several forms, some of which are publicly available. eg: https://docs.validator.com.au/nocc/02.10/appendix-a.html#file-header-record . In many places there are zeroes that need to be (and were being) preserved.

All the specification have a version string that is `\d\d.\d\d`, like `03.30` which when parsed to 3.3 and doesn't match its intended format. Another field must be `\d{4}` with its missing value defined as `missing: 0003`; similarly 3 doesn't match the intended format.

For now, we've pinned YAML::XS@0.86 in our builds as a short term fix.

**Comment by perlpunk** [comment]
> > An unquoted scalar is a number according to the YAML spec
> 
> Double checking I've understood the spec correctly; that arrises from the extended list of regexs from the core schema?

That's a not so trivial question. YAML::XS doesn't implement the YAML 1.2 Core schema, but also not the 1.1 types. It only loads things as numbers for which the perl function `looks_like_number` returns true.

And so far it used `SvIV_please` on those, resulting in scalars which have both the IV or NV and the PV flag on.

> > Can you tell us more about the use case that relies on this behaviour
> 
> We have 18+ years of data collection specifications that use YAML docs as their primary source. They get processed into several forms, some of which are publicly available. eg: https://docs.validator.com.au/nocc/02.10/appendix-a.html#file-header-record . In many places there are zeroes that need to be (and were being) preserved.
> 
> All the specification have a version string that is `\d\d.\d\d`, like `03.30` which when parsed to 3.3 and doesn't match its intended format. Another field must be `\d{4}` with its missing value defined as `missing: 0003`; similarly 3 doesn't match the intended format.

So in general if you want to have leading or trailing zeroes preserved, the correct thing to do is quote it in your YAML files. If you are ever using a different YAML parser (in perl or a different language) then you will lose those zeroes.

I can imagine that it would be possible to still preserve the PV with the zeroes and turn off POK at the same time, but that is something I have to try out, and will cost me time.

Reverting the change will result in others not using YAML::XS, and making it configurable is also quite some work.

@ingydotnet any comments from you?

**Comment by ingydotnet** [comment]
The only thing that comes to mind immediately is that we add an option to do it either way.

**Comment by perlpunk** [comment]
btw, here are several YAML modules listed with their behaviour on various strings: https://perlpunk.github.io/YAML-PP-p5/schema-examples.html

**Comment by perlpunk** [comment]
For the record, I tested:
```
SvIV_please(scalar);
SvPOK_off(scalar);
```
instead of using SvIOK_only/SvNOK_only and it had the same effect of losing the zeroes.

**Comment by veryrusty** [comment]
Thanks for the comments and further investigation @perlpunk & @ingydotnet .

After kicking it over with the team here, we will work back through our yaml docs and quote the values that should be treated as strings. As much of a PITA that will be, it'll be less than burdening you with ongoing maintenance of a option/switch to allow either behaviour.

As such, this can be closed.

**Comment by perlpunk** [comment]
@veryrusty 
Apparently there are others with similar problems, so we are probably going to revert the change and figure out what to do next.

**Comment by ingydotnet** [comment]
#112

**Comment by karenetheridge** [comment]
Note that you can use single quotes in the YAML data to indicate that the value should be interpreted as a string, not a number. e.g. compare these two:

```
$ perl -MData::Dumper -MYAML::XS -wle'print Dumper(Load("---\n- 0.3\n- 0.30\n- 03.30\n"))'
$VAR1 = [
          '0.3',
          '0.3',
          '3.3'
        ];

$ perl -MData::Dumper -MYAML::XS -wle'print Dumper(Load("---\n- 0.3\n- \x{27}0.30\x{27}\n- \x{27}03.30\x{27}\n"))'
$VAR1 = [
          '0.3',
          '0.30',
          '03.30'
        ];
```

(`\x{27}` used here rather than a literal `'` only because it's a shell oneliner)

**Comment by karenetheridge** [comment]
As an alternative solution, perhaps a "interpret all numbers as strings" config value could be added, but it should not default to on, because it contradicts the yaml spec.

**Comment by karenetheridge** [comment]
Are there any additional thoughts for what to do about this?

**Comment by perlpunk** [comment]
I can imagine adding an option that offers a way around this behaviour, but not as a default. The current behaviour must stay the default, otherwise it would break too much; that is what we know know.
But I have a hundred small things on my plate; and when I think of doing something for YAML::XS I always hesitate because I have to workaround Zilla::Dist, and over the years I got really tired of it. It doubles the amount of work for an XS module, because you cannot just call `prove -l`. If we had Dist::Zilla here, I would do it.
So if @ingydotnet would allow such an option, PRs are welcome, but I don't feel like doing it.

**Comment by karenetheridge** [comment]
@ingy would you consider allowing a small development-tailored Makefile.PL in the repository for XS dists? Moose (etc) does this to allow a developer to still do `perl Makefile.PL; make test` in the repository.

**Comment by karenetheridge** [comment]
@ingydotnet  would you consider allowing a small development-tailored Makefile.PL in the repository for XS dists? Moose (etc) does this to allow a developer to still do `perl Makefile.PL; make test` in the repository.

**Comment by perlpunk** [comment]
> @ingydotnet would you consider allowing a small development-tailored Makefile.PL in the repository for XS dists? Moose (etc) does this to allow a developer to still do `perl Makefile.PL; make test` in the repository.

That will be solved by #117

**Comment by karenetheridge** [comment]
The underlying issue (that the 0.87 release was addressing, via https://github.com/ingydotnet/yaml-libyaml-pm/pull/107) is still open, that both POK and IOK/NOK are being set on the decoded value, which is wrong. Pick one of these, and don't set both, as it makes it impossible for code (e.g. other serializers such as Cpanel::JSON::XS) to know whether the value was "originally" a number or a string.

Does https://github.com/ingydotnet/yaml-libyaml-pm/pull/110 resolve this? If so can it please be merged/released?

**Comment by perlpunk** [comment]
The problem is that the current behaviour is something that people rely on. It's kind of a "preserve exact string that a number was loaded from" feature by accident, because perl has dual vars.
So this could only be fixed by adding a new option, to ensure backwards compatibility.
We wanted to get rid of the global variables for the options for a long time, but didn't have time.

I have recently started to work on an object oriented variant of YAML::XS, possibly released in the same CPAN distribution, that will
* Load and Dump values according to the YAML 1.2 Core schema
* Only set IV or NV for numbers
* Set options in the constructor, per instance instead of gloal vars

Maybe I can work on it more during SUSE's hackweek in november.

**Comment by karenetheridge** [comment]
> The problem is that the current behaviour is something that people rely on.

That's the "interpret values like `0030.1` as strings, not numbers, yes? This can still be achieved by _only_ setting the PV flag on the scalar, not also the NV flag.  I don't have an issue with how values are parsed as strings vs integers - only that a numeric scalar is being set with _both_ PV and NV (or IV) at the same time, which should never be necessary.

**Comment by perlpunk** [comment]
Ok, let's go over the `03.30` example like described in this issue at the top.
`03.30` in YAML (in that case YAML::XS behaves correctly according to the 1.2 schema) is a number and should be loaded as one, but currently it has IV, NV and PV set, if loaded via YAML::XS.
Yes, that's not necessary for a number, but that's because [YAML::XS uses `SvIV_please()`](https://github.com/ingydotnet/yaml-libyaml-pm/blob/d23c3b1f85438a87b26c7ca00024225b7381f595/LibYAML/perl_libyaml.c#L486) to convert the SV.
The number is `3.3`, the leading zeroes are dropped. That's what's in the NV slot.
The PV has `03.30`, the exact original string.

A side effect of this is, that if you dump this again, you will get the exact `03.30` as output again:

```
% perl -wE'
use YAML::XS;
use Devel::Peek ();
my $yaml = "03.30";  
my $data = Load $yaml;
Devel::Peek::Dump $data;
my $dump = Dump $data;
say $dump;'
SV = PVNV(0x5581e68d1310) at 0x5581e68f0c88
  REFCNT = 1
  FLAGS = (NOK,POK,pIOK,pNOK,pPOK)
  IV = 3
  NV = 3.3
  PV = 0x5581e68faa30 "03.30"\0
  CUR = 5
  LEN = 10
--- 03.30
```

I don't know if that was originally intended or not, so let's call it an accidental feature.
Some people are relying on that.

**Comment by karenetheridge** [comment]
> I don't know if that was originally intended or not, so let's call it an accidental feature.

Okay, what if we only set the PV flag if the stringified form of the number is different from what's stored in the string slot?  That is, only these special backcompat cases will be dualvars, and all other numbers will just be numbers.

That is, `03.30` will parse as PV = "03.30", NV = 3.3 as before, but `3.14` will parse as NV = 3.14 with no PV at all. That would also save on memory, too.

Arguably, the spec conformance change should be re-applied, with the backcompat support still available with a configuration option for those who were relying on that behaviour.

**Comment by perlpunk** [comment]
> Okay, what if we only set the PV flag if the stringified form of the number is different from what's stored in the string slot? That is, only these special backcompat cases will be dualvars, and all other numbers will just be numbers.
> 
> That is, `03.30` will parse as PV = "03.30", NV = 3.3 as before, but `3.14` will parse as NV = 3.14 with no PV at all. That would also save on memory, too.

I'm surprised that would help in your case. `03.30` and `003` should still be detected as numbers, e.g. in a validator. That's your use case, right?
But yeah, that would be possible, once I figure out the correct XS functions to use.

> Arguably, the spec conformance change should be re-applied, with the backcompat support still available with a configuration option for those who were relying on that behaviour.

Well, even with that change, YAML::XS is not spec compliant, as it doesn't implement the YAML 1.2 core schema correctly. (And also not YAML 1.1).

That's why I started the new OOP variant, which would be spec compliant from the start, and also get rid of some of the older options (which still can be used when using the old implementation). And it would load numbers as numbers only.
And I would avoid having to implement another option for the old loader.

**Comment by karenetheridge** [comment]
> I'm surprised that would help in your case. 03.30 and 003 should still be detected as numbers, e.g. in a validator. That's your use case, right?

Yes, but "is a string" or "is a number" validation fails for dualvars, since it's not possible to tell if the value is really intended to be a string or a number, and different serializers might treat this value differently, so for example when decoding yaml and then re-encoding as json, or writing to a database, we might guess the wrong type.


```
perl -wE'
use YAML::XS;
use Devel::Peek ();
my $yaml = "3.30";         # note no leading 0
my $data = Load $yaml;
Devel::Peek::Dump $data;
my $dump = Dump $data;
say $dump;
use Cpanel::JSON::XS;
say encode_json({foo => $data});
say encode_json({foo => 0+$data})'
SV = PVNV(0x159809470) at 0x15b0148f0
  REFCNT = 1
  FLAGS = (NOK,POK,pIOK,pNOK,pPOK)
  IV = 3
  NV = 3.2999999999999998
  PV = 0x6000000e81e0 "3.30"\0
  CUR = 4
  LEN = 16
--- 3.30

{"foo":"3.30"}
{"foo":3.3}
```

**Comment by perlpunk** [comment]
Ok, I see that Cpanel::JSON::XS and also JSON::PP treat those values differently if the PV value differs from the IV/NV value.
I guess I should also change YAML::PP to do the same.
However, the right thing to do here is to load all these values as numbers without a PV:
```
- 3
- 3.1
- 03.30
```
If I implement your suggested change to do it only for the first two, then the third value wouldn't be recognized as a number, but it *is* a number (it's `3.3`).
The correct JSON serialization for this YAML is `[3, 3.1, 3.3]`.
So it's got to be a new option, or the new loader interface.

**Comment by karenetheridge** [comment]
> If I implement your suggested change to do it only for the first two, then the third value wouldn't be recognized as a number, but it is a number (it's 3.3).

Yes, but it's still better than the current situation, where _all_ numbers are not recognized as numbers because they have a PV slot. And it preserves the behaviour that the user who created this ticket is depending on.

(If you're building a test suite for these scenarios, I see that the yaml `3e1` also behaves oddly -- PV="3e1", NV=30. I would assume that the spec says that should also just be the number 30.)

**Comment by perlpunk** [comment]
I created an experimental release in my fork: https://github.com/perlpunk/yaml-libyaml-pm/releases/tag/untagged-a3392fdae261ba363b5b
Besides bringing an OOP interface and implementing the YAML 1.2 Core Schema, it only sets the IV/NV flag for numbers now.
It still needs some cleanup and error handling, but should be good for playing around.
